tools/ttysnoop.py

```diff
@@ -97,9 +144,15 @@ def usage():
     if args.ebpf:
         exit()
 
+bpf_text = bpf_text.replace('USER_DATASIZE', '%s' % args.datasize)
+bpf_text = bpf_text.replace('USER_DATACOUNT', '%s' % args.datacount)
+
 # initialize BPF
 b = BPF(text=bpf_text)
 
+if BPF.get_kprobe_functions(b'tty_write'):
+    b.attach_kprobe(event="tty_write", fn_name="trace_tty_write")
```

Comment:
This won't work. I tried on a 5.2 kernel and see:
```
207: kprobe  name tty_write  tag 41976baf6bf5cfd7  gpl
        loaded_at 2021-04-19T16:56:28-0700  uid 0
        xlated 3488B  jited 2092B  memlock 4096B  map_ids 100,99
        btf_id 45
208: kprobe  name trace_tty_write  tag ab95056fe92735a8  gpl
        loaded_at 2021-04-19T16:56:29-0700  uid 0
        xlated 3672B  jited 2204B  memlock 4096B  map_ids 100,99
        btf_id 45

// run bpftool perf
...
pid 120451  fd 7: prog_id 207  kprobe  func tty_write  offset 0
pid 120451  fd 9: prog_id 208  kprobe  func tty_write  offset 0
```

Basically we have 2 kprobe programs attached to tty_write. There are two problems here for old kernel: first really old kernel like 4.4 may be 4.9 not supporting kprobe attaching two bpf programs, and second I see your code still use bpf_perf_event_output with map value which is not supported for 4.4 either.

I checked the commit changing the signature and some surrounding commits,
```
9bb48c82aced07698a2d08ee0f1475a6c4f6b266 tty: implement write_iter
dd78b0c483e33225e0e0782b0ed887129b00f956 tty: implement read_iter
3b830a9c34d5897be07176ce4e6f2d75e2c8cfd7 tty: convert tty_ldisc_ops 'read()' function to take a kernel pointer
```
and cannot find useful info (visitable header macro/struct_type change) etc. which the script can be used to detect whether the signature is changed. So in this case, I think use version is OK, I don't think people backports tty stuff, which should be extremely rare.

Your first version of code looks largely fine. But please keep the original code for kernel versions < 5.11 and the new code >= 5.11. This way, the old kernel including 4.4 will be still supported. Please also add some comments in the code around the place where you check kernel versions to mention we do kernel version check since we could not find a reliable way to detect the precise point where the kernel signature is changed. Thanks!

Reply:
ok, will post new version, thanks